Add client-side form binding reference (#1299)
vaadin/flow-and-components-documentation: documentation/client-side-forms/appendix-client-side-form-binding-reference.asciidoc

Technically, every model instance represents either a key of a parent model, or the value of the Binder itself. The model constructors receive the following arguments:

- `parent: ModelParent<T>` — the parent object reference: either the parent model instance, or the `Binder` instance
- `key: keyof any` — the key in the parent object to get the value from / set the value to, could be `string` when the parent is an object or `number` when the parent is an array
- `...validators: ReadonlyArray<Validator<T>>` — the default validators, typically exported from Java JSR-380 annotations

=== AbstractModel

All models subclass from the `AbstractModel<T>` TypeScript class, where the `T` type argument refers to the value type.

AbstractModel, as well as any subclass, define a static method `static createEmptyValue(): T`, that returns the empty value of the subject model type:

[source, typescript]
----
const emptyPerson: Person = PersonModel.createEmptyValue();
console.log(emptyPerson); // {"fullName": ""}
----

As with any JavaScript object, `AbstractModel` has `toString(): string` and `valueOf(): T` instance methods, that are handy for template expressions:

> [soruce, html]

Reviewer: typo
```suggestion
[source, html]
```

  ${model.middleName.toString()}
  ${model.lastName.toString()}
</p>

<p>
  Cost:
   ${model.quantity.valueOf() * model.price.valueOf()}
</p>
----

=== Primitive models

There are the builtin models that represent the common primitive field types:

|===
| Type | Value type `T` | Empty value (`Model.createEmptyValue()` result)

| `StringModel` | `string` | `''`
| `NumberModel` | `number` | `0`
| `BooleanModel`  | `boolean` | `false`
|===

Primitive models subclass `PrimitiveModel<T>`. Primitive models are leaf, i. e., do not have nested field keys.

=== Object models

`ObjectModel<T>` is a common superclass for all generated bean models.

There is no API added in the `ObjectModel<T>` over `AbstractModel<T>` itself.

The generated subclasses of `ObjectModel<T>` define the type argument constrains and the default type, as well as list the bean fields as in Java , e. g. for the following Java bean:

.IdEntity.java
[source, java]
----
public class IdEntity {
    private String idString;

    public String getIdString() {
        return idString;
    }

    public void setIdString(String idString) {
        this.idString = idString;
    }
}
----

.Person.java

> [source,java]

Reviewer: Minor consistency issue. All other blocks in this file have a space after the comma here. (looks like it's quite inconsistent in the whole repository too, but would be good to at least try to keep some kind of consistency for new content.)
```suggestion
[source, java]
```

Reviewer 2: Yeah, it’s very inconsistent. I would prefer to have the space after the comma as well.

----
import javax.validation.constraints.NotNull;
import javax.validation.constraints.NotEmpty;

public class Person extends IdString {
    @NotEmpty(message = "Cannot be empty")
    private String fullName;

    public String getFullName() {
        return fullName;
    }

    public void setFullName(String fullName) {
        this.fullName = fullName;
    }
}
----

The following TypeScript interfaces are generated for type-checking endpoints:

.IdEntity.ts

> [souces, typescript]

Reviewer: typo
```suggestion
[source, typescript]
```

 ></vaadin-text-field>
----

Depending on the type of the bound component, the field directive selects a strategy, which defines how exacly the states above are applied on the component, e. g., which attributes and properties of the element are used.

[NOTE]
You can find more information on field strategy customization in the <<appendix-web-component-field-strategy#, Appendix: Using a Web Component Field>> article.

The field directive supports Vaadin components and HTML input elements. While Vaadin components have support for all the states, for HTML input elements the `invalid`, `required` and `errorMessage` are not displayed in the bound component. As a workaround, you can bind those manually in the template:

[source, html]
----
<label for="fullName">
  Full name
  ${binder.for(binder.model.fullName).required ? '*' : ''}
</label>
<input id="fullName" ...="${field(binder.model.fullName)}"/><br/>
${
  binder.for(binder.model.fullName).invalid
    ? html`
      <strong>
        ${binder.for(binder.model.fullName).errors[0]}
      </strong>`
    : ''
}
----

[NOTE]
See also: <<appendix-vaadin-components.asciidoc#, Appendix: Binding Data to Vaadin Components>>

== Model

> The model is a type-safe reference to a value stored in the binder. Models describe the structure of the data, and are used to read and modify current form values and other related properties and states.

Reviewer 2: Style (minor): a general recommendation is that each sentence is written on a separate line. Makes it easier to review changes and change the order of sentences when editing.
```suggestion
The model is a type-safe reference to a value stored in the binder. 
Models describe the structure of the data, and are used to read and modify current form values and other related properties and states.
```

Vaadin automatically generates Model classes for server-side endpoints from Java beans. Usually there is no need to define models manually.

Technically, every model instance represents either a key of a parent model, or the value of the Binder itself (e. g., the form data object).

[NOTE]
====
The `Binder` constructor receives the form model class, and takes care of creating model instances. More on that in the `Binder` section below.
====

//The model constructors receive and store the following arguments:
//
//- `parent: ModelParent<T>` — the parent object reference: either the parent model instance, or the `Binder` instance
//- `key: keyof any` — the key in the parent object to get the value from / set the value to, could be `string` when the parent is an object or `number` when the parent is an array
//- `...validators: ReadonlyArray<Validator<T>>` — the default validators, typically exported from Java JSR-380 annotations

> === Abstract model

Reviewer 2: ```suggestion
=== Abstract Model
```

All the models subclass from the `AbstractModel<T>` TypeScript class, where the `T` type argument refers to the value type.

`AbstractModel<T>`, as well as any subclass, define a static method `static createEmptyValue(): T`, that returns the empty value of the subject model type:

[source, typescript]
----
const emptyPerson: Person = PersonModel.createEmptyValue();
console.log(emptyPerson); // {"fullName": ""}
----

As with any JavaScript object, `AbstractModel` has `toString(): string` and `valueOf(): T` instance methods, that are handy for template expressions:

[source, html]
----
<p>
  Full name:
  ${model.firstName.toString()}
  ${model.middleName.toString()}
  ${model.lastName.toString()}
</p>

<p>
  Cost:
   ${model.quantity.valueOf() * model.price.valueOf()}
</p>
----

[NOTE]
====
The actual data referenced with every model instance, such as the current value, the validators, the required flag, and the validation state, is available on the corresponding `BinderNode` (see below).
====

> === Primitive models

Reviewer 2: ```suggestion
=== Primitive Models
```

> There are the builtin models that represent the common primitive field types:

Reviewer: ```suggestion
These are the builtin models that represent the common primitive field types:
```

Reviewer 2: Also, should be ”built-in”, I believe.

Reviewer: Good catch, I updated my suggestion.

|===
| Type | Value type `T` | Empty value (`Model.createEmptyValue()` result)

| `StringModel` | `string` | `''`
| `NumberModel` | `number` | `0`
| `BooleanModel`  | `boolean` | `false`
|===

Primitive models subclass `PrimitiveModel<T>`. Primitive models are leaf, i. e., do not have nested field keys.

> === Object models

Reviewer 2: ```suggestion
=== Object Models
```

> Primitive models subclass `PrimitiveModel<T>`. Primitive models are leaf, i. e., do not have nested field keys.

Reviewer: (minor) I think this would be easier to read and understand. I wouldn't use "subclass" as a verb since it can be confused for a substantive.
```suggestion
Primitive models extend `PrimitiveModel<T>`. Primitive models are leaf nodes of the data structure, that is, they do not have nested field keys.
```

Typically, the primitive editable values of the form are grouped in objects. `ObjectModel<T>` accommodates that. `ObjectModel<T>` is also a common superclass for all the models generated from Java beans.

The subclasses of `ObjectModel<T>` define the type argument constrains and the default type. In addition, the subclasses list all the public properties, following the shape of the described object.

For example, for the following Java bean:

.IdEntity.java
[source, java]
----
public class IdEntity {
    private String idString;

    public String getIdString() {
        return idString;
    }

    public void setIdString(String idString) {
        this.idString = idString;
    }
}
----

.Person.java
[source,java]
----

> import javax.validation.constraints.NotNull;

Reviewer: This import is not used in the example so I think this line should be removed?

import javax.validation.constraints.NotEmpty;

> public class Person extends IdString {

Reviewer: I guess this should be:
```suggestion
public class Person extends IdEntity {
```
?

}
----

.Person.ts
[source, typescript]
----
import IdEntity from './IdEnity';

export default interface Person extends IdEntity {
  fullName: string;
}
----

And the following models are generated for client-side form binding:

.IdEntityModel.ts
[source, typescript]
----
import IdEntity from './IdEntity';

export default class IdEntityModel<T extends IdEntity = IdEntity> extends ObjectModel<T> {
  static createEmptyValue: () => IdEntity;
  readonly idString = new StringModel(this, 'idString');
}
----

.PersonModel.ts
[source, typescript]
----
import IdEntityModel from './IdEntityModel';

import Person from './Person';

export default class PersonModel<T extends Person = Person> extends IdEntityModel<T> {
  static createEmptyValue: () => Person;
  readonly fullName = new StringModel(this, 'fullName', new NotEmpty({message: 'Cannot be empty'}));
}
----

[IMPORTANT]
====
To avoid naming collisions with user-defined object model fields, the builtin models and model superclasses do not have any public instance properties or methods, aside form the `toString` and `valueOf` exceptions mentioned above.
====

[NOTE]
====
The properties of object models are intentionally read-only.
====

> === Array model

Reviewer 2: ```suggestion
=== Array Model
```

> ====

Reviewer 2: Not sure if all these admonition blocks are necessary (across the entire article, not these two in particular)? It might make it harder to read the article if they are used too frequently.

Reviewer 2: @magi42, do you have any guidelines when admonitions should be used?

Author: I’ve inlined some of them here.

Reviewer 2: Okay, great! Good if you could identify some that don’t benefit from being called out more. I don’t really have any concrete guidance for these, just that too much of those will make it hard to read :)

Array models are iterable, iterating yields binder nodes for entries:

[source, html]
----
${repeat(this.binder.model.people, personBinder => html`
  <div>
    <vaadin-text-field
     label="Full name"
     ...="${field(personBinder.model.fullName)}"
     ></vaadin-text-field>
    <strong>Full name:</strong>
    ${personBinder.value.fullName}
  </div>
`)}
----

The array entries are not available for indexing with bracket notation (`[]`).

== Binder

The `Binder` class acts as a controller. It hosts the form top-level model instance, the default and the current value, performs validation on value changes, as well as updates the form view component to display any changes of the values and the validation state.

The form model is instantiated by the `Binder`.

The `Binder` constructor arguments are:

`context: Element`::
The form view component instance to update.
`Model: ModelConstructor<T, M>`::
The constructor (the class reference) of the form model. The `Binder` instantiates the top-level model and
`config?: BinderConfiguration<T>`::
The options object.
  `onChange?: (oldValue?: T) => void`:::

>   The callback that updates the form view, by default uses `context.requestUpdate()` is used.

Reviewer: "uses ... is used" doesn't make sense.
```suggestion
  The callback that updates the form view, by default uses `context.requestUpdate()`.
```

  `onSubmit?: (value: T) => Promise<T | void>`:::
  The endpoint for submitting the form data into.

The `Binder` has the following instance properties:

`model: M`::
The form model, the top-level model instance created by the `Binder`.
`value: T`::
The current value of the form, two-way bound to the field components.
`defaultValue: T`::
The initial value of the form, before any fields are edited by the user.
`readonly validating: boolean`::
True when there is an ongoing validation.
`readonly submitting: boolean`::
True if the form was submitted, but the submit promise is not resolved yet.

The `Binder` instance methods are:

`reset(defaultValue?: T): void`::
Reset the form to the default value. If the argument is given, sets the default value property to the argument value first.
`clear(): void`::
Sets the form to empty value, as defined in the Model.
`getFieldStrategy(element: any): FieldStrategy`::
Determines and returns the `field` directive strategy for the bound element. Override to customise the binding strategy for a component.

The `Binder` extends `BinderNode`, see the inherited properties and methods below.

> == Binder nodes

Reviewer 2: ```suggestion
== Binder Nodes
```

> The `BinderNode<T, M>` class provides the from binding related APIs with respect to a particular model instance.

Reviewer: ```suggestion
The `BinderNode<T, M>` class provides the form binding related APIs with respect to a particular model instance.
```
